pkg/psalabelsyncer/sccrolecache.go

```diff
@@ -0,0 +1,440 @@
+	if binding := r.clusterRoleBinding; binding != nil {
+		return binding.RoleRef
+	}
+	return r.roleBinding.RoleRef
+}
+
+func (r *roleBindingObj) Subjects() []rbacv1.Subject {
+	if binding := r.clusterRoleBinding; binding != nil {
+		return binding.Subjects
+	}
+	return r.roleBinding.Subjects
+}
+
+func (r *roleBindingObj) Namespace() string {
+	if r.clusterRoleBinding != nil {
+		return ""
+	}
+	return r.roleBinding.Namespace
+}
+
+// AppliesToNS returns true if:
+// - the wrapped object is a cluster role binding
+// - the namespace of a wrapped rolebinding matches the namespace supplied in `ns`
+func (r *roleBindingObj) AppliesToNS(ns string) bool {
+	if r.clusterRoleBinding != nil {
+		return true
+	}
+	return ns == r.roleBinding.Namespace
+}
+
+// roleObj helps to handle roles and clusterroles in a generic manner
+type roleObj struct {
+	role        *rbacv1.Role
+	clusterRole *rbacv1.ClusterRole
+}
+
+func newRoleObj(obj interface{}) (*roleObj, error) {
+	switch r := obj.(type) {
+	case *rbacv1.ClusterRole:
+		return &roleObj{
+			clusterRole: r,
+		}, nil
+	case *rbacv1.Role:
+		return &roleObj{
+			role: r,
+		}, nil
+	case *roleObj:
+		return r, nil
+	default:
+		return nil, fmt.Errorf("the object is neither a Role, nor a ClusterRole: %T", obj)
+	}
+}
+
+func (r *roleObj) Rules() []rbacv1.PolicyRule {
+	if role := r.clusterRole; role != nil {
+		return role.Rules
+	}
+	return r.role.Rules
+}
+
+func (r *roleObj) Name() string {
+	if role := r.clusterRole; role != nil {
+		return role.Name
+	}
+	return r.role.Name
+}
+
+func (r *roleObj) Namespace() string {
+	if role := r.clusterRole; role != nil {
+		return role.Namespace
+	}
+	return r.role.Namespace
+}
+
+// BySAIndexKeys is a cache.IndexFunc indexing function that shall be used on
+// rolebinding and clusterrolebinding informer caches.
+// It retrieves the subjects of the incoming object and if there are SA, SA groups
+// or the system:authenticated group subjects, these will all be returned as a slice
+// of strings to create an index for the SA or SA group.
+func BySAIndexKeys(obj interface{}) ([]string, error) {
+	roleBinding, err := newRoleBindingObj(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccounts := []string{}
+	for _, subject := range roleBinding.Subjects() {
+		if subject.APIGroup == "" && subject.Kind == "ServiceAccount" {
+			serviceAccounts = append(serviceAccounts, serviceaccount.MakeUsername(subject.Namespace, subject.Name))
+		} else if subject.APIGroup == rbacv1.GroupName && subject.Kind == "Group" &&
+			(subject.Name == serviceaccount.AllServiceAccountsGroup ||
+				subject.Name == user.AllAuthenticated ||
+				strings.HasPrefix(subject.Name, serviceaccount.ServiceAccountGroupPrefix)) {
+			serviceAccounts = append(serviceAccounts, subject.Name)
+		}
+	}
+
+	return serviceAccounts, nil
+}
+
+func NewSAToSCCCache(rbacInformers rbacv1informers.Interface, sccInfomer securityv1informers.SecurityContextConstraintsInformer) *SAToSCCCache {
```

Comment:
This looks like a potential scale problem to me.  The for loops in here are pretty deeply nested.  If this becomes a bottleneck, how would you make it an order of magnitude faster.  The information is present to fully invert this logic.

Reply:
The logic expects minimal changes in terms of SCC creation, which is currently the most expensive operation as the role cache gets reinitialized. That also means that the initialization of the cache on the start is going to be quite expensive - each SCC addition means `(n_SCCs * (n_roles * n_role_rules))` complexity. This could perhaps be improved by the cache dealing with only the SCC that was changed, and therefore not reinitializing the whole cache.

As for the roles and clusterroles - these are handled one-by-one and cached. The computation complexity there is not great - `(n_SCCs * n_role_rules)` - but I think that in this case it's either us or the kube-apiserver doing the computation.

As for the memory complexity, the role cache is just a `string->[]strings` where the string slice are the names of the SCCs. This means that the cache will of course grow with role and SCC numbers, but the number of SCCs should hopefully never be too great and so memory growth here should be rather linear.

Reply:
The cost for adding an SCC has been improved to `(n_roles * n_role_rules)`

Reply:
@s-urbaniak proposed to write a benchmark test as a follow-up